### DISCUSSION TITLE
Trouble using MCP23017 and encoders : value increments

### DISCUSSION THREAD

**Discussion by Su24nn3** [tttapa/Control-Surface#766] [Q&A] [answered]
Hi !

I have tried using a MCP23017 (on a module like this : https://fr.aliexpress.com/item/1005001360480752.html) with a Teensy 4.1, 2 encoders. 
I have used this example : https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/db3/MCP23017-Encoders-Interrupts_8ino-example.html
I get values that just increment, no matter which encoder I turn, no matter which direction I turn the encoder.
I get this result on the serial monitor (turning different directions on both encoders) : 
```
[0] 2
[1] 2
[0] 4
[1] 4
[0] 6
[1] 6
[0] 8
[1] 8
[1] 10
[0] 10
[0] 12
[1] 12
[1] 13
[0] 14
[1] 14
[0] 16
[1] 16
```

Any idea how to debug ?

Thanks :)

**Comment by tttapa** [comment, marked as answer]
This is usually caused by a wiring issue (e.g. by swapping the common pin with one of the signal pins).

**Reply by Su24nn3** [reply]
Thanks for the reply
I already checked the wiring and that doesn't seem to be the case. 
Also I had already tested my encoders separately in case the pinout wasn't standard but they were normal.

**Reply by tttapa** [reply]
From the output you posted, it seems like the encoder pins are off by one: encoder 1 should be connected to the first and second pin, but it seems like it might be connected to the second and third pin instead. Could that be the issue?

**Reply by Su24nn3** [reply]
You mean pins on the MCP23017 side ? GPA0, etc. ? If so I checked and everything is alright there : encoder 1 on A0 - A1, encoder 2 on A2  -A3.
Could it have something to do with the module I'm using ? Do you use MCP23017 on breakout boards or just the chip + resistors ?

**Reply by Su24nn3** [reply]
Ok, I found a wiring issue ! I hadn't completely pulled out my resistors from a previous test 😅

I'm still getting values that increment 4 times for each small turn on the knob.

Like this : 
```
[1] -1
[1] -2
[1] -3
[1] -4
[0] 5
[0] 6
[0] 7
[0] 8
```
Is this normal behaviour ?

**Reply by tttapa** [reply]
Glad to hear you found it.

Yes this is normal, each physical click of the encoder corresponds one complete cycle in the output, and one cycle consists of 4 distinct states (2 outputs with 1 bit each, so four possible combinations).  
If you want, you can just divide the result by four.

**Reply by Su24nn3** [reply]
Ok great, thank you :)
